### PR TITLE
6️⃣ Change default `LookupIpStrategy` back to `Ipv4AndIpv6`

### PR DESCRIPTION
**Pull request by database64128**
- #300 doesn't make any sense. #301 seems like an arbitrary change to me.
- Reverts #301 6ec1ec3.

### CODE REVIEW

**Comment by djc** [comment]
Just dismissing previous issues/PRs as "doesn't make any sense" or "arbitrary change" is not a great way to make your argument. The point in #300 that the behavior for `Ipv4AndIpv6` is non-deterministic makes a lot of sense to me, so just reverting this outright does not seem like the best way forward.

**Comment by database64128** [comment]
> The point in #300 that the behavior for `Ipv4AndIpv6` is non-deterministic makes a lot of sense to me

Can you elaborate? I don't think the current bahavior of `Ipv4AndIpv6` can cause any problem for common applications. You can always set it to `Ipv4Only` if your application can't handle IPv6 addresses, which really isn't a good look in 2020.

**Comment by bluejekyll** [comment]
I would love to make the default to be ipv4 and/or ipv6, but we can't do that right now. The issue at the heart of #300 and #301 is twofold, one is that when ipv6 is not supported (which is still true for a lot of people behind various ISP configurations), and the current API for trust-dns-resolver for getting results is not appropriate for this use case as it is not a streaming async API.

Let me describe why these are both important, as it exists right now, the API will block and wait for both ipv4 and ipv6 responses. In some DNS configurations, the ipv6 response never comes, causing the ipv4 response to be delayed until the ipv6 response times out. There were many discussions, some may be across more issues than the two linked and possibly in the discord channels.

In order for us to consider making the default to be ipv4 and/or ipv6 be the default again, we first need to fix the API to be streaming. (The current API is also broken in the mDNS case for the same reason, as there are times when multiple response are expected in mDNS). I would love to make happy-eyeballs in some form the default again, but first we need to change the API to be more appropriate for that.

In short, I think your statement stands as the best response to this:

> You can always set it to Ipv4Only if your application can't handle IPv6 addresses, which really isn't a good look in 2020.

You can flip that around. Also, I'll say, it might be 2020, but that doesn't mean IPv6 is now better supported on home routers than in 2017. For example, here is a very popular gigabit ISP in my area with statements going back to 2011 about support for IPv6 coming, but as of yet still not done: https://forums.sonic.net/viewtopic.php?f=10&t=10735

**Comment by database64128** [comment]
Thank you for the detailed explanation.

**Comment by nirvana-msu** [comment]
This is an old issue but I'd like to state that I believe premise behind #300 (which led to change the default) is wrong. It is responsibility of whatever client is making requests to ensure it finds/picks a reliable way to route the request (such as by using happy eyeballs or whatever else), and not a responsibility of the DNS resolution library. The fact that the client simply picked first record in #300 (which happened to be IPv6) and failed to connect is not an issue with `trust-dns` in any way.

The default, ideally, should be to return both IPv4 and IPv6 records. See e.g. https://github.com/seanmonstar/reqwest/issues/2049 for the kind of unexpected issues it leads to - e.g. IPv6-only hosts failing to connect by default when enabling `trust-dns` feature for reqwest.

The concern with latency and requiring a streming API is a valid one, of course.

**Comment by bluejekyll** [comment]
@nirvana-msu , can you open a new issue for this? We can discuss options (like when binding to ipv6 use ipv6 lookups?

**Comment by SRv6d** [comment]
> This is an old issue but I'd like to state that I believe premise behind #300 (which led to change the default) is wrong. It is responsibility of whatever client is making requests to ensure it finds/picks a reliable way to route the request (such as by using happy eyeballs or whatever else), and not a responsibility of the DNS resolution library. The fact that the client simply picked first record in #300 (which happened to be IPv6) and failed to connect is not an issue with `trust-dns` in any way.
> 
> The default, ideally, should be to return both IPv4 and IPv6 records. See e.g. [seanmonstar/reqwest#2049](https://github.com/seanmonstar/reqwest/issues/2049) for the kind of unexpected issues it leads to - e.g. IPv6-only hosts failing to connect by default when enabling `trust-dns` feature for reqwest.
> 
> The concern with latency and requiring a streming API is a valid one, of course.

Strongly agree with this. If I understand the scope of this library correctly, it is not to be a drop in replacement for obsolete C APIs that center around IPv4 (due to their origins being 40 odd years ago) like [`gethostbyname`](https://man7.org/linux/man-pages/man3/gethostbyname.3.html), but to be a general purpose DNS library. As such (I think) it should not concern itself with limitations to be handled by downstream code, but rather just provide the data as-is.  

`getaddrinfo`, the replacement for `gethostbyname` which was standardized some 20 years ago resolves both IPv4 and IPv6 while returning IPv6 first on most platforms. 

Preferring IPv6 over IPv4 if available is the defacto standard on modern platforms and not doing so will cause problems on networks that cannot provide native IPv4 connectivity and must do so through through translation mechanisms, which is increasingly common in Mobile Networks as well as with ISPs that have come up within the last 10 years and do not have large IPv4 allocations they can fall back on. 
Using a default that **will be broken on certain networks** to support downstream code that does not correctly handle the full IP standard (as it has been for the last 20 years) is a bad choice IMO, especially since rust is a fairly new language, so "it needs to be like this for legacy reasons" should not be a valid argument. 
As an example, [Apple does not allow apps that prefer IPv4 over IPv6 in it's App Store since 2016](https://developer.apple.com/support/ipv6/).

**Comment by bluejekyll** [comment]
@SRv6d, and everyone interested in this, let's please move this discussion to an open issue where we might be able to come up with solutions to the current issues. We currently have this open issue, #1507, which I think is fundementally about the same problem so maybe we could move it there?

I definitely want to come up with a solution here, but as I described up thread, I think we need to improve the async design of the Resolver in order to do that. For example, this old issue is related to what we need to change, #383. I'm all for leaving the old getaddrinfo and other poor designs in the past, but it will take work. So I think we need to come up with some workarounds in the meantime as reworking the interior types to be properly streaming will be quite a bit of work.

**Comment by SRv6d** [comment]
@bluejekyll Is https://github.com/hickory-dns/hickory-dns/issues/383 really required to be implemented before? 
In my opinion, an implementation where both `A` and `AAAA` queries are done at (roughly) the same time and both results return only after both queries either are successful or fail/timeout would be perfectly fine, and https://github.com/hickory-dns/hickory-dns/issues/1507 more of an (optional) optimization. Since both requests will typically be answered by the same DNS server using the same address family, cases where the `AAAA` query times out while the `A` query does not should be extremely rare.

**Comment by bluejekyll** [comment]
We had this as the default in fact, querying both in parallel. The experience as I remember it was really bad in some cases as IPv6 queries wouldn’t just fail, but instead always timeout. That’s why we ended up settling on what we have now.

One thing that I don’t think we’ve investigated is trying to do something a little more interesting when using IPv6 transports. That is, we could reverse the default based on if the underlying interface is IPv6 and do IPv6thenIPv4, that might be a decent compromise until there’s a decent streaming API available?

**Comment by SRv6d** [comment]
I can't really imagine why that would happen, do you have an example issue? If we send both queries to the same resolver using the same IP family, for the `AAAA` query to time out while the `A` does not the recursive/authoritative resolver would need to have a bug where an `AAAA` query times out, while other types do not. I have not yet seen this happen with common implementations, even if they are very old(15+ years). Another possibility would be that the `AAAA` response is truncated and truncation is not properly implemented or breaks due to the network. I don't think `AAAA` being truncated while `A` is not for the same name is common but I have not data for this. Nonetheless, this seems more likely than the previous scenario. 

> One thing that I don’t think we’ve investigated is trying to do something a little more interesting when using IPv6 transports. That is, we could reverse the default based on if the underlying interface is IPv6 and do IPv6thenIPv4, that might be a decent compromise until there’s a decent streaming API available?

I'm somewhat confused by this, what does the transport have to do with the query type? Are you suggesting to query only `A` if the network only appears to support IPv4 and `AAAA` as well as `A` if the network appears to support IPv4 as well as IPv6?

**Comment by bluejekyll** [comment]
I'm not finding a reference to the issues I'm worried about from the past, I guess it could have been a discussion elsewhere. The concern I have at this point is that it's very possible that we'd end up breaking a lot of consumers of the API, which is the basis for the original issue that this stems from.

> Are you suggesting to query only A if the network only appears to support IPv4 and AAAA as well as A if the network appears to support IPv4 as well as IPv6?

basically. If you are sending a request to an ipv6 addressed name server, then I would suggest we'd switch the default to be Ipv6thenIpv4, if you are communicating with a name server over ipv4 then the default would be ipv4thenipv6. This would prioritize one over the other in those cases.

The thing that we really need is some way to test this in the real world without breaking everyone who uses the library, and that's a tough thing to do, which is why I'm so hesitant on changing it. People can always experiment with all of the different settings on the request, as that's a config option.

**Comment by SRv6d** [comment]
> I'm not finding a reference to the issues I'm worried about from the past, I guess it could have been a discussion elsewhere. The concern I have at this point is that it's very possible that we'd end up breaking a lot of consumers of the API, which is the basis for the original issue that this stems from.

Totally understand not wanting to break consumers, although https://github.com/hickory-dns/hickory-dns/issues/300 might've also been a breaking change for consumers. Maybe a deprecation or some sort of warning could be used for a while before, should the default be changed again?

> basically. If you are sending a request to an ipv6 addressed name server, then I would suggest we'd switch the default to be Ipv6thenIpv4, if you are communicating with a name server over ipv4 then the default would be ipv4thenipv6. This would prioritize one over the other in those cases.

I think this is a bad idea since the underlying IP family used to make DNS requests should have nothing to do with the query type. Querying the resolver using IPv4 could be because the resolver only supports IPv4 (while the client supports v6), or the resolver might even support v6 while the user is not aware of it. It is not correct to assume the client does not support IPv6 based on the IP family used to connect to the resolver.

While I personally think choosing an address family is the responsibility of the client when using an external DNS library, a great middle ground that should not break consumers would be to implement a `getaddrinfo` like default.

On most platforms `getaddrinfo` works by querying for A(IPv4) and AAAA(IPv6), regardless of the underlying family for resolution, and then sorting the results by relevance to the consumer. This means that on hosts `getaddrinfo` deems IPv6 capable AAAA results are returned first, followed by A. If no IPv6 is available, it does the opposite, first returning A. The consumer can then try the first result and move on to the next should the first one fail.

**Comment by bluejekyll** [comment]
> This means that on hosts getaddrinfo deems IPv6 capable AAAA results are returned first, followed by A. If no IPv6 is available, it does the opposite, first returning A.

Do you know the logic here? This could be a decent middle ground if it's easy to determine. I don't think IPv6 interfaces are necessarily routable, so just the presence probably isn't enough?

**Comment by djc** [comment]
> Totally understand not wanting to break consumers, although https://github.com/hickory-dns/hickory-dns/issues/300 might've also been a breaking change for consumers.

That change happened in 2017 -- there were likely a *lot* fewer consumers at the time.

**Comment by SRv6d** [comment]
> Do you know the logic here? This could be a decent middle ground if it's easy to determine. I don't think IPv6 interfaces are necessarily routable, so just the presence probably isn't enough?

I haven't looked at the code, but I think in essence the logic is IPv4 connectivity is assumed if there is an IPv4 address on any interface except the loopback, while IPv6 connectivity is assumed if there is an IPv6 address (excluding link local) on any interface except the loopback. Before implementing this it would make sense thought to look at the `getaddrinfo` implementation in more detail since it **might** make sense to exclude some interfaces like tunnel interfaces for example if `getaddrinfo` does this.
